javalib/src/main/scala/java/util/regex/README.md

```diff
@@ -0,0 +1,322 @@
+* atomic groups `(?>𝑋)`,
+* possessive quantifiers `𝑋*+`, `𝑋++` and `𝑋?+`,
+* the `\A`, `\Z` and `\z` boundary matchers,
+* the `\R` expression,
+* embedded quotations with `\Q` and `\E`, both outside and inside character classes.
+
+In addition, among others, the following features have the correct semantics from Java, even though their corresponding feature in JavaScript has different semantics:
+
+* the `^` and `$` boundary matchers with the `MULTILINE` flag (when the latter is supported),
+* the predefined character classes `\h`, `\s`, `\v`, `\w` and their negated variants, respecting the `UNICODE_CHARACTER_CLASS` flag,
+* the `\b` and `\B` boundary matchers, respecting the `UNICODE_CHARACTER_CLASS` flag,
+* the internal format of `\p{𝘯𝘢𝘮𝘦}` character classes, including the `\p{java𝘔𝘦𝘵𝘩𝘰𝘥𝘕𝘢𝘮𝘦}` classes,
+* octal escapes and control escapes.
+
+### Guarantees
+
+If a feature is not supported, a `PatternSyntaxException` is thrown at the time of `Pattern.compile()`.
+
+If `Pattern.compile()` succeeds, the regex is guaranteed to behave exactly like on the JVM, *except* for capturing groups within repeated segments (both for their back references and subsequent calls to `group`, `start` and `end`):
+
+* on the JVM, a capturing group always captures whatever substring was successfully matched last by *that* group during the processing of the regex:
+  - even if it was in a previous iteration of a repeated segment and the last iteration did not have a match for that group, or
+  - if it was during a later iteration of a repeated segment that was subsequently backtracked;
+* in JS, capturing groups within repeated segments always capture what was matched (or not) during the last iteration that was eventually kept.
+
+The behavior of JavaScript is more "functional", whereas that of the JVM is more "imperative".
+This imperative nature is also reflected in the `hitEnd()` and `requireEnd()` methods of `Matcher`, which we do not support (they don't link).
+
+The behavior of the JVM does not appear to specified, and is questionable.
+There are several open issues that argue it is buggy:
+
+* https://bugs.openjdk.java.net/browse/JDK-8027747
+* https://bugs.openjdk.java.net/browse/JDK-8187083
+* https://bugs.openjdk.java.net/browse/JDK-8187080
+* https://bugs.openjdk.java.net/browse/JDK-8187082
+
+Therefore, it seems wise to keep the JavaScript behavior, and not try to replicate the JVM behavior at great cost (if that is even possible within our constrains).
+
+## Implementation strategy
+
+Java regexes are compiled to JS regexes in one pass, using a recursive descent approach.
+There is a state variable `pIndex` which indicates the position inside the original `pattern`.
+Compilation methods parse a subexpression at `pIndex`, advance `pIndex` past what they parsed, and return the result of the compilation.
+
+Parsing is always done at the code point level, and not at the individual `Char` level.
```

Comment:
~~I think this might need expanding (or maybe a footnote): It is unclear what happens if there are lone surrogates.~~

Seems like the "handling unicode section" below explains this, maybe reference it in a paren?